### DISCUSSION TITLE
LoadLibraryEx() and FreeLibrary()

### DISCUSSION THREAD

**Discussion by Arebusf21** [microsoft/CsWin32#839] [Q&A] [answered]
PInvoke.LoadLibraryEx returns a FreeLibrarySafeHandle:
`internal static unsafe FreeLibrarySafeHandle LoadLibraryEx(string lpLibFileName, SafeHandle hFile, winmdroot.System.LibraryLoader.LOAD_LIBRARY_FLAGS dwFlags)`

PInvoke.FreeLibrary() requires a HINSTANCE parameter:
`internal static extern winmdroot.Foundation.BOOL FreeLibrary(winmdroot.Foundation.HINSTANCE hLibModule);`

How can I get a winmdroot.Foundation.HINSTANCE from a FreeLibrarySafeHandle?
or am I missing something?

also;

The 2nd parameter to LoadLibraryEx should be null, according to the documentation. But null throws an exception:
`throw new ArgumentNullException(nameof(hFile));
					winmdroot.Foundation.HINSTANCE __result = PInvoke.LoadLibraryEx(lpLibFileNameLocal, hFileLocal, dwFlags);
					return new FreeLibrarySafeHandle(__result, ownsHandle: true);`

Using VS2022, v17.4.3, .Net 7.0, Windows Target OS 10.0.20348 Console App. and Microsoft.Windows.CsWin32 0.2.164-beta

See Program.cs
[EnumerateResources.zip](https://github.com/microsoft/CsWin32/files/10333876/EnumerateResources.zip)

**Comment by AArnott** [comment, marked as answer]
`LoadLibraryEx` is emitted twice: once with a `SafeHandle` return type and once with `HINSTANCE` as its return type. It is only necessary to call `FreeLibrary` when you use the `HINSTANCE` returning overload. When you use the `SafeHandle` overload, you should just dispose of the `SafeHandle` when you're done with it, and it will call `FreeLibrary` on your behalf. 

The null-throwing issue you found is a bug in the metadata, which I've reported for you in https://github.com/microsoft/win32metadata/issues/1421.
